lib/cmdline.c

```diff
@@ -1167,42 +1145,130 @@ static bool rm_cmd_set_cmdline(RmCfg *cfg, int argc, char **argv) {
     return true;
 }
 
-static bool rm_cmd_set_paths(RmSession *session, char **paths) {
-    bool is_prefd = false;
-    bool all_paths_valid = true;
-    bool stdin_paths_preferred = false;
-
-    RmCfg *cfg = session->cfg;
-
-    /* Check the directory to be valid */
-    for(int i = 0; paths && paths[i]; ++i) {
-        if(strcmp(paths[i], "-") == 0) {
-            cfg->read_stdin = true;
+typedef struct rm_cmd_set_paths_vars {
+    RmCfg *const cfg;
+    char **const paths;
+    unsigned int index;
+    bool is_prefd;
+    bool read_stdin;
+    bool stdin_paths_preferred;
+    const bool null_separated;
+    bool all_paths_valid;
+} rm_cmd_set_paths_vars;
+
+static INLINE
+void rm_cmd_set_paths_from_cmdline(
+    rm_cmd_set_paths_vars *const v,
+    const bool replay
+) {
+    g_assert(v);
+    g_assert(v->paths);
+
+    for(char *path, **list = v->paths; (path = *list); ++list) {
+        if(path[0] == '-' && path[1] == 0) {
+            v->read_stdin = true;
             /* remember whether to treat stdin paths as preferred paths */
-            stdin_paths_preferred = is_prefd;
-        } else if(strcmp(paths[i], "//") == 0) {
+            v->stdin_paths_preferred = v->is_prefd;
+        } else if(path[0] == '/' && path[1] == '/' && path[2] == 0) {
             /* the '//' separator separates non-preferred paths from preferred */
-            is_prefd = !is_prefd;
+            v->is_prefd = !v->is_prefd;
         } else {
-            all_paths_valid &= rm_cfg_add_path(cfg, is_prefd, paths[i]);
+            v->all_paths_valid &= rm_cfg_prepend_path(
+                v->cfg, path, v->index++, replay, v->is_prefd
+            );
+        }
+        g_free(path);
+    }
+    g_free(v->paths);
+}
+
+static INLINE
+bool rm_cmd_set_paths_from_stdin(
+    rm_cmd_set_paths_vars *const v,
+    const bool replay
+) {
+    g_assert(v);
+    g_assert(v->cfg);
+
+    RmCfg *const cfg = v->cfg;
+    const bool is_prefd = v->stdin_paths_preferred;
+
+    char delim = v->null_separated ? 0 : '\n';
+
+    size_t buf_len;
+    char *path_buf = malloc(buf_len = PATH_MAX);
```

Comment:
Please revert that. It's easy enough to miss that assignment. I generally prefer direct initialization together with declaration.